cudapoa/samples/sample_cudapoa.cpp

```diff
@@ -124,22 +126,127 @@ void process_batch(Batch* batch, bool msa, bool print)
+        max_lengths[i] = max_read_length;
+    }
+
+    // create histogram based on number of max POAs
+    size_t num_bins = 20;
+    std::vector<size_t> bins_frequency(num_bins, 0);             // represents the count of windows that fall within the corresponding range
+    std::vector<size_t> bins_max_length(num_bins, 0);            // represents the length of the window with maximum sequence length in the bin
+    std::vector<size_t> bins_num_reads(num_bins, 0);             // represents the number of reads in the window with maximum sequence length in the bin
+    std::vector<size_t> bins_ranges(num_bins, 1);                // represents maximum POAs
+    std::vector<std::vector<size_t>> bins_window_list(num_bins); // list of windows that are added to each bin
+
+    for (size_t j = 1; j < num_bins; j++)
+    {
+        bins_ranges[j] = bins_ranges[j - 1] * 2;
+    }
+
+    // go through all windows and keep track of the bin they fit
+    for (size_t i = 0; i < num_windows; i++)
+    {
+        for (size_t j = 0; j < num_bins; j++)
+        {
+            if (max_poas[i] <= bins_ranges[j] || j == num_bins - 1)
+            {
+                bins_frequency[j]++;
+                bins_window_list[j].push_back(i);
+                if (bins_max_length[j] < max_lengths[i])
+                {
+                    bins_max_length[j] = max_lengths[i];
+                    bins_num_reads[j]  = windows[i].size();
+                }
+                break;
+            }
+        }
+    }
+
+    // a bin in range N means a batch made based on this bean, can launch up to N POAs. If the sum of bins frequency of higher ranges
+    // is smaller than N, they can all fit in batch N and no need to create extra batches.
+    // For example. consider the following:
+    //
+    // bins_ranges      1        2       4       8       16      32      64      128     256     512
+    // bins frequency   0        0       0       0       0       0       10      51      0       0
+    // bins width       0        0       0       0       0       0       5120    3604    0       0
+    //
+    // note that bin_ranges represent max POAs. This means larger bin ranges follow with smaller corresponding max lengths (windows with shorter reads)
+    // In the example above, to process 10 windows that fall within bin range 64, we need to create one batch. This batch can process up to 64 windows of
+    // max length 5120 or smaller. This means all the windows in bin range 128 can also be processed with the same batch and no need to launch an extra batch
+
+    size_t remaining_windows = num_windows;
+    for (size_t j = 0; j < num_bins; j++)
+    {
+        if (bins_frequency[j] > 0)
+        {
+            list_of_batch_sizes.emplace_back(bins_max_length[j], bins_num_reads[j]);
+            list_of_windows_per_batch.push_back(bins_window_list[j]);
+            remaining_windows -= bins_frequency[j];
+            if (bins_ranges[j] > remaining_windows)
```

Comment:
This will possibly result in working only at the tail end of the histogram. Is that intended?
```
 // bins_ranges       1        2       4       8       16      32      64      128     256     512
 // bins frequency    0        0       2       2       0       0       10      51      0       0
 // bins width            0        0       X       Y(<X)       0       0       5120    3604    0       0
```

Reply:
That's correct. It was intentional, since in practice, or at least in the bonito tests cases, there were fewer windows that needed the larger bins (those with smaller max POA), and majority of the windows would be toward the tail. So I did not put any effort on collapsing intermediate bins into each other. Toward the end, each bin has a higher capacity (larger max POA) meaning corresponding batch can fit larger number of windows launched in parallel, also as mentioned earlier, most of windows fall within these ranges, therefore it was written to simply collapse the last bins into one.

Reply:
@atadkase trying with some other datasets, I noticed it is needed to revise the strategy as you pointed, to merge intermediate bins whenever possible as well. I will revise it to be more general.

Reply:
changed.